### DISCUSSION TITLE
Rationale behind weird API design of Symfony Mailer

### DISCUSSION THREAD

**Discussion by severak** [symfony/symfony#45432] [Q&A] [answered]
### Description

Could someone please explain me rationale behind weird design of Symfony Mailer API?

For context:

I was tasked with upgrading some old websites to newer version of Symfony Framework. Those old websites are using custom mail transport. It's just very simple code which just takes email from Swiftmailer and CURLs it into some proprietary API.

I struggled while trying to port it to Symfony Mailer. I followed [this tutorial](https://albertmoreno.dev/posts/creating-custom-symfony-mailer-transports/) only to find out what abomination `\Symfony\Component\Mailer\Transport\TransportInterface` is.
	
When you are trying to implement `send` method you got two arguments:

- `\Symfony\Component\Mailer\Envelope` object which contains sender and recipient addresses
- `\Symfony\Component\Mime\RawMessage` object which contains... who knows what - it's not possible to guess it from design of that object

`\Symfony\Component\Mime\RawMessage` has absolute zero of documentation, it only possible to find out that it's something which is:

- serializable
- iterable
- possible to stringify
- and it has to be valid (but `ensureValidity` method is not implemented)

After some poking in the mailer code I discovered that `RawMessage` is extended first by `\Symfony\Component\Mime\Message` and then by `\Symfony\Component\Mime\Email`, both of which are much more usable than `RawMessage` itself. My question is: **WHY**?

- **Why** there is whole object hiearchy if there is not any other descendand of `RawMessage` than `\Symfony\Component\Mime\Message`?
- **Why** there is not a single line of documentation in `RawMessage`? (for humans not linters)
- **Why** `RawMessage` even exists if it's not usable on itself?

I am tagging @fabpot himself as he is listed as author of these files and he is probably one who designed these weird interfaces.

I understand this is not feature request nor bug report. This is mostly something like documentation issue (because there is no documentation about `RawMessage` at all) but I think that weird API design is more issue of software itself not issue of the documentation.

Any explanation or rationale is welcomed.

---

Code in question:

- [Symfony\Component\Mailer\Transport\TransportInterface](https://github.com/symfony/symfony/blob/44db49fb8f6d70c4341120a3f8f1b1a1a0dfa014/src/Symfony/Component/Mailer/Transport/TransportInterface.php)
- [Symfony\Component\Mime\RawMessage](https://github.com/symfony/symfony/blob/44db49fb8f6d70c4341120a3f8f1b1a1a0dfa014/src/Symfony/Component/Mime/RawMessage.php)

**Comment by Wirone** [comment, marked as answer]
@severak You may be interested in symfony/symfony#33394 (especially [my comment](https://github.com/symfony/symfony/issues/33394#issuecomment-791295930)) 😉

**Reply by severak** [reply]
Thanks for sharing it.

I solved my issue simply by NOT USING new Symfony mail interface. I simply used old code and everything worked.
